### DISCUSSION TITLE
How to properly query array columns with filter?

### DISCUSSION THREAD

**Discussion by orenlande** [supabase/supabase#16588] [Questions] [answered]
Hey, I'm struggling to figure out what's the right method to filter the following column:
```
database
... | ... | faction_ids: int4[] | ... | ...
... | ... | [3,5]               | ... | ...
```

supbase js:
```
const { data: decks, error } = await supabase
      .from("decks")
      .select(
        "*, cards(*, types!cards_type_id_fkey(type_code)), deck_views(views_count)"
      )
      .filter("faction_ids", "in", "(3)") <=== throws error
      .order("created_at", { ascending: false })
      .range(offset, limit);
```
 I keep getting the following error:
```
{
    "code": "42704",
    "details": null,
    "hint": null,
    "message": "could not find array type for data type integer[]"
}
```

I guess it's something small to adjust, but just can't understand exactly. Thanks for your help!

**Comment by GaryAustin1** [comment, marked as answer]
.in() uses an array or list of values to compare against a column with a single value (not array column).
.contains(),.containedBy() are two of the main filters on array columns.
https://supabase.com/docs/reference/javascript/contains
https://supabase.com/docs/reference/javascript/containedby

**Reply by orenlande** [reply]
awesome!
and say I want to make a dynamic query using the `.filters` method as the following, how can I achieve the similar results? Maybe using `cs`? I tried and didn't work out

```
    const filters = [];

    if (queries.faction_ids && queries.faction_ids.length) {
      filters.push(["cs", "faction_ids", queries.faction_ids]);
    }

    const { data: decks, error } = await filters.reduce(
      (acc, [filter, ...args]) => {
          return acc[filter](...args);
      },
      supabase
        .from("decks")
        .select(
          "*, cards(*, types!cards_type_id_fkey(type_code)), deck_views(views_count)"
        )
        .order("created_at", { ascending: false })
    );
```

**Reply by GaryAustin1** [reply]
I have no idea what your code is doing there...
![image](https://github.com/supabase/supabase/assets/54564956/42337065-345e-4fc0-a29c-8e1d3ecea065)
Shows using .filter with cs.

**Reply by orenlande** [reply]
The idea is to build a query dynamically [I took the cue from this thread](https://github.com/orgs/supabase/discussions/787)

Something like this:
```
const filtersMap = {}

if (someConditionMet = true) {
  filtersMap["newFilter"] = [3,4,6] // factions ID
}

// execute supabase query here
```

My purpose is to build a query based on the filters passed in, and by using `filter()` I understand I lose the opportunity to build a dynamic query. Unless I misunderstood and there a better way to build it dynamically, I'd love to know how.

**Reply by GaryAustin1** [reply]
I just do conditional chaining.  There might be ways like you are showing, but I've never attempted that.
![image](https://github.com/supabase/supabase/assets/54564956/3272e8dd-19b3-4701-8f28-67b9e6a32e75)

**Reply by orenlande** [reply]
Ok yeah, it makes much more sense. Hard to find this information on the docs. Sorry!

Hope it's ok I'm adding another question on this thread - is there a way to check PARTIAL inclusion? something like `ANY` in SQL language. aka if the array in db is `[2,3]` and I'm querying for all rows which include the number 2. Using the contains or containedBy checks for full inclusion as far I understand, and won't return rows that contain more elements.

**Reply by GaryAustin1** [reply]
contains should do what you want.  The database array has to include all the items in your search array.  So 2,3 does include 2.  If you had 2,4 as your compare array then 2,3 would not contain both 2 and 4 so would not return.

Edit:  I marked the original question as solved.  You really should ask new questions as few wanting to help tend to look in questions with lots of posts that no longer relate to the original topic.
